doc/developer/design/20230802_explain_running_dataflows.md

```diff
@@ -0,0 +1,255 @@
+Even though our optimizer is deterministic, this will only provide results consistent with the dataflow that is actually running while the environment context that guides the optimizer remains unchanged.
+For example, if the user does:
+
+```sql
+-- The definition of V optimized with an empty set of indexes.
+CREATE MATERIALIZED VIEW V AS (SELECT * FROM T where a = 5);
+-- A new index on T(a) is created.
+CREATE T_a INDEX ON T(a);
+-- The explain is produced from an optimization run that assumes that an index on T(a) exists.
+EXPLAIN MATERIALIZED VIEW V;
+```
+
+The `EXPLAIN` output will indicate that we are using `T_a` for the point lookup, while in practice this will not be the case because the index was created _after_ installing the dataflow that maintains `V`.
+
+This design doc proposes a sequence of changes to the catalog and our `EXPLAIN` handling code that would allow us to change the behavior of
+
+```sql
+EXPLAIN [OPTIMIZED|PHYSICAL] PLAN FOR MATERIALIZED VIEW
+```
+
+and implement similar behavior for
+
+```sql
+EXPLAIN [OPTIMIZED|PHYSICAL] PLAN FOR INDEX
+```
+
+# Explanation
+[explanation]: #explanation
+
+At the moment, `sequence_explain_plan` constructs its output using information derived from optimizations done at the time when `sequence_explain_plan` runs.
+
+At the high level, the API that constructs an `EXPLAIN` output for a given plan looks like that:
+
+```rust
+let context = ExplainContext {
+    config,
+    humanizer,
+    used_indexes,
+    finishing,
+    duration,
+};
+
+Explainable::new(&mut plan).explain(format, context)
+```
+
+We need to
+(1) analyze which values in the above code snippet represent information that should reflect the state of the world when the dataflow was originally created, and
+(2) suggest mechanisms to make these values available for subsequent `EXPLAIN <catalog item>` calls.
+
+In general, we can persist the required data
+(a) in-memory, as extensions of the `CatalogItem` objects that are maintained as entries in the `CatalogState`, or
+(b) durably, as extensions of the `SerializedCatalogItem` structures that represent the part of the catalog state that survives restarts.
+
+Since at the moment every `CatalogItem` is fully re-optimized upon an `environmentd` restart in the catalog re-hydration phase, the current proposal focuses on option (a).
+However, note that if we introduce plan pinning or similar features in the future, we might need to revisit this decision.
+
+The comments in the following snippet summarize the high-level implementation plan:
+
+```rust
+let context = ExplainContext {
+    config: &config,     // Derived from the enclosing `EXPLAIN` statement plan.
+    humanizer: &catalog, // Passing the current catalog state.
+    used_indexes,        // Passing a `CatalogItem` field.
```

Comment:
Is this a collection of `GlobalId`s?

Reply:
Yes, at the moment it is:

```rust
pub struct UsedIndexes(Vec<GlobalId>);
```

In the future this might change to disambiguate indexes used for full scans vs point lookups, but if we have this collection we can re-construct the rest lazily.

Reply:
Sorry for the unrelated question, but why do we need this separate list of indexes when we have a plan? I suppose that means the plan doesn't specify `Get(index_id)` directly but references `used_indexes` somehow instead?

Reply:
>  I suppose that means the plan doesn't specify `Get(index_id)` directly but references used_indexes somehow instead?

Exactly. I wish we had `IdxFetch(index_id)` and `IdxLookup(index_id, <expr>)` nodes, but we don't. We need to do a a pass at the plan contextualized with an (immutable for the duration of the pass) set of indexes in order to decide which indexes are actually going to be picked up at rendering time.